pkg/controllers/nodeclaim/underutilization/controller.go

```diff
@@ -0,0 +1,450 @@
+		// if the nodeclaim doesn't have the underutilized status condition, it's not part of
+		// an active consolidation.
+		cond := s.NodeClaim.StatusConditions().GetCondition(v1beta1.Underutilized)
+		if !cond.IsTrue() {
+			return
+		}
+		// If the nodepool isn't found, we assume it's deleted or cluster state hasn't seen it yet.
+		// If it was deleted, we rely on the termination controller to clean this up
+		np, found := nodePools[s.Node.Labels[v1beta1.NodePoolLabelKey]]
+		if !found {
+			return
+		}
+		cmd := commands[cond.Message]
+		// Add the command ID in idempotently
+		cmd.CommandID = cond.Message
+		cmd.Candidates = append(cmd.Candidates, s)
+		// Use the larger consolidateAfter when batching
+		cmd.ConsolidateAfter = lo.Max([]time.Duration{cmd.ConsolidateAfter, *np.Spec.Disruption.ConsolidateAfter.Duration})
+		commands[cmd.CommandID] = cmd
+	})
+	return commands
+}
+
+// Returns true if the candidates should be unmarked if used to validate existing commands.
+// Returns a nil error if the candidates are valid.
+// This intentionally does a quick check rather than doing a scheduling simulation.
+// Candidates are considered invalid if:
+// 1. They're not disruptable from cluster state's perspective
+// 2. ConsolidateAfter is disabled
+// 3. Is NotEmpty, but it's NodePool only allows WhenEmpty consolidation
+// 4. NodePool Disruption Budgets do not allow any disruption
+func (c *Controller) validateCandidates(ctx context.Context, candidates ...*state.StateNode) error {
+	newCandidates := lo.Filter(candidates, func(candidate *state.StateNode, _ int) bool {
+		return candidate.IsDisruptable(ctx, c.kubeClient, c.recorder) != nil
+	})
+	if len(newCandidates) != len(candidates) {
+		return fmt.Errorf("detected non-disruptable candidates")
+	}
+
+	nodePools := c.cluster.NodePools()
+	// Group candidates by their NodePool so that we can calculate allowed disruptions
+	nodePoolsToCandidates := lo.GroupBy(newCandidates, func(candidate *state.StateNode) string {
+		return candidate.Labels()[v1beta1.NodePoolLabelKey]
+	})
+
+	for _, cn := range newCandidates {
+		// If the NodePool isn't using WhenEmpty, checking if it's an invalid candidate can't be done quickly.
+		if nodePools[cn.Labels()[v1beta1.NodePoolLabelKey]].Spec.Disruption.ConsolidationPolicy != v1beta1.ConsolidationPolicyWhenEmpty {
+			continue
+		}
+		pods, err := cn.ReschedulablePods(ctx, c.kubeClient)
+		if err != nil {
+			return fmt.Errorf("getting candidate pods, %w", err)
+		}
+		if len(pods) > 0 {
+			return fmt.Errorf("found non-empty node with NodePool consolidationPolicy set to %q", v1beta1.ConsolidationPolicyWhenEmpty)
+		}
+	}
+
+	// Cross reference each node to its nodepool:
+	// 1. check if consolidation is enabled
+	// 2. check if node disruption budgets allowed disruptions is 0
+	for npName := range nodePoolsToCandidates {
+		nodePool := nodePools[npName]
+		if nodePool.Spec.Disruption.ConsolidateAfter.Duration == nil {
```

Comment:
If consolidation isn't enabled, can't this check be done quickly too?

Reply:
Or maybe I'm misunderstanding the way that you are structuring this. From the look of this, seems like you are checking 3 invariants, but (maybe it's code structure), this wasn't immediately clear to me on the first read. It seemed like you had some checks that you specifically wanted to do for emptiness and then some others that you wanted to do for WhenUnderutilized